src/app/clusters/commodity-tariff-server/CommodityTariffAttrsDataMgmt.h

```diff
@@ -0,0 +1,827 @@
+        {
+            mUpdateState = UpdateState::kValidated;
+        }
+        return err;
+    }
+
+    /**
+     * @brief Commits the validated new value
+     *
+     * @note Performs the following actions:
+     *       1. Checks if value actually changed
+     *       2. Cleans up old value if needed
+     *       3. Updates stored value
+     *       4. Invokes callback if provided
+     *       5. Resets to kIdle state
+     *
+     * @pre Must be in kValidated state (after UpdateBegin())
+     * @post Always transitions to kIdle state
+     */
+    void UpdateCommit()
+    {
+        /* Skip if the attribute object has no new attached data */
+        if (mUpdateState == UpdateState::kIdle)
+        {
+            return;
+        }
+
+        assert(mUpdateState == UpdateState::kValidated);
+
+        if (HasChanged())
+        {
+            if (mHoldState == StorageState::kHold)
+            {
+                CleanupValue(mValue);
+            }
+
+            mValue     = mNewValue;
+            mHoldState = StorageState::kHold;
+
+            if (mAuxCb)
+            {
+                mAuxCb(mAttrId, mAuxData);
+            }
+
+            mUpdateState = UpdateState::kUpdated;
+        }
+
+        UpdateEnd();
+    }
+
+    /**
+     * @brief Aborts the current update process
+     *
+     * @note Performs the following actions:
+     *       1. Discards pending new value
+     *       2. Cleans up if needed
+     *       3. Resets to kIdle state
+     *
+     * @post Always transitions to kIdle state
+     */
+    void UpdateEnd()
+    {
+        /* Skip if the attribute object has no new attached data */
+        if (mUpdateState == UpdateState::kIdle)
+        {
+            return;
+        }
+        /* In cases if new data was initialized/activated but has not assigned to Value storage */
+        if (mUpdateState != UpdateState::kUpdated)
+        {
+            CleanupValue(mNewValue);
+            //mNewValue    = mValue;
+            mUpdateState = UpdateState::kIdle;
+        }
+    }
+
+    void Cleanup() { CleanupValue(mValue); }
+
+protected:
+    T & mValue;             // Reference to the applied value storage
+    T mNewValue;            // Value for updating
+    void * mAuxData;        // Pointer to an auxiliary data which can be used in some method implementations
+    const uint32_t mAttrId;
+    void (*mAuxCb)(uint32_t, void *);
+
+    StorageState mHoldState  = StorageState::kEmpty;
+    UpdateState mUpdateState = UpdateState::kIdle;
+
+    bool is_changed = false;
+
+    /**
+     * @brief Validate a new value
+     * @param newValue The value to validate
+     * @return CHIP_NO_ERROR if valid, an err code otherwise
+     *
+     * @note Derived classes should override for custom validation
+     */
+    virtual CHIP_ERROR Validate(const valueType & aValue) const { return CHIP_NO_ERROR; }
+
+    virtual bool CompareStructValue(const PayloadType & source, const PayloadType & destination) const
```

Comment:
Needs documentation.  It's very non-obvious that "Compare" means "return true if not same".